src/process/extract/IndexReader.cpp

```diff
@@ -0,0 +1,186 @@
+    }
+
+    if (this->readHeader.dictionariesAreCompressed) {
+        // We could only check, if there is at least one entry.
+    } else if (0 != (fileSize - headerSize) % sizeOfIndexEntry) {
+        addErrorMessage("Cannot read index file, there is a mismatch between stored index version and content size.");
+        indexFile->close();
+        return false;
+    }
+
+    if (this->readHeader.numberOfEntries > 0) { // Easy case, it is stored.
+        this->indicesLeft = this->readHeader.numberOfEntries;
+    } else if(!this->readHeader.dictionariesAreCompressed){
+        this->indicesLeft = (fileSize - headerSize) / sizeOfIndexEntry;
+    }
+
+    if(this->indicesLeft == 0) {
+        addErrorMessage("Could not properly determine the amount of indices in this file.");
+        indexFile->close();
+        return false;
+    }
+
+    this->indicesCount = indicesLeft;
+
+    this->readerIsOpen = true;
+
+    return readerIsOpen;
+}
+
+IndexHeader IndexReader::readIndexHeader() {
+    IndexHeader header;
+    indexFile->read((Bytef*)&header, sizeof(IndexHeader));
+
+    this->readHeader = header;
+    headerWasRead = true;
+    return header;
+}
+
+vector<shared_ptr<IndexEntry>> IndexReader::readIndexFile() {
+    vector<shared_ptr<IndexEntry>> convertedLines;
+    if (!tryOpenAndReadHeader()) {
+        addErrorMessage("Could not read index file due to one or more errors during file open.");
+        return convertedLines;
+    }
+
+    // Read in and convert a specific header version to an IndexEntry vector
+    if (this->readHeader.indexWriterVersion == 1) {
+        auto entries = readIndexFileV1();
+        for (auto const &entry : entries) {
+            convertedLines.emplace_back(entry->toIndexEntry());
+        }
+    } // We do not need an else branch, a version range check is applied earlier.
+
+    return convertedLines;
+}
+
+vector<shared_ptr<IndexEntryV1>> IndexReader::readIndexFileV1() {
+    vector<shared_ptr<IndexEntryV1>> convertedLines;
+    // Initialize an empty index entry to make following steps a bit easier.
+    while (indicesLeft > 0) {
+        auto indexEntry = readIndexEntryV1();
+        if (indexEntry) {
+            convertedLines.emplace_back(indexEntry);
+        }
+    }
+    return convertedLines;
+}
+
+shared_ptr<IndexEntry> IndexReader::readIndexEntry() {
+    if (!tryOpenAndReadHeader()) {
+        addErrorMessage("Could not read index file due to one or more errors during file open.");
+        return shared_ptr<IndexEntry>(nullptr);
+    }
+
+    // Read in and convert a specific header version to an IndexEntry vector
+    if (this->readHeader.indexWriterVersion == 1) {
+        return readIndexEntryV1()->toIndexEntry();
+    } // We do not need an else branch, a version range check is applied earlier.
+}
+
+shared_ptr<IndexEntryV1> IndexReader::readIndexEntryV1() {
+    if (!readerIsOpen) {
+        addErrorMessage("You have to open the IndexReader instance first with tryOpenAndReadHeader()");
```

Comment:
This error condition is actually a programming bug, or not? Nice to have an error for such code-usage related issues. Maybe an exception would be better?